### PR TITLE
runtime/memprof.c: store configuration fields on the major heap

### PR DESCRIPTION
**Pull request by gasche**
I have found two different fixes for #14300, this one and #14306.
I wonder if @NickBarnes has a preference between either approaches.

---

The documentation says:

```
 * However, there are some structures shared between domains. The main
 * such structure is the profile configuration object on the Caml
 * heap. The only field written in this object is the status field,
 * used to communicate between domains sharing the profile, when a
 * profile is stopped or discarded. This field is inspected or set
 * atomically by the `Status` and `Set_status` macros.
```

But in fact the `lambda` and `1log1ml` fields are boxed float values, and boxed float values get moved around by the GC during promotion, resulting in a write to the `1log1ml` field that results in a TSan race report.

To avoid this race, we ensure that both float values are allocated on the major heap directly, so that the corresponding fields in the `config` record are not mutated by the GC. (Well, they can in fact be moved around during compaction.)

### CODE REVIEW

**Comment by gasche** [comment]
I find this fix more robust than the one in #14305, but it partially reverts a change done by @NickBarnes earlier, so we should understand if the removal of the second check is an issue or not in practice. It might have a performance impact (in the case with a neg-infinity sampling rate, the pre-PR code uses the fast path more often), or it may be a convenience impact (one way to stop Memprof was to pass a neglectible `lambda` value,  I guesss.)

**Comment by OlivierNicole** [comment]
I think that I’m arriving to the same conclusion as you regarding the cause of #14300: that the racy read is the one performed by `Min_lambda(config)`, which is the reading of a boxed float. The racy write is one by the major heap allocator as part of a promotion.

I have yet to understand why this write in the free list ends up racing with another domain.

> To avoid this race, we ensure that both float values are allocated on the major heap directly, so that the corresponding fields in the `config` record are not mutated by the GC. (Well, they can in fact be moved around during compaction.)

Indeed they can, so my feeling is that we probably need to address the problem higher up in the chain of causality here—a read in the runtime is not supposed to race with a write in the allocation code.

**Comment by gasche** [comment]
The problem is that `memprof` stores some of its runtime state on the OCaml heap. Most of the runtime state (data that is manipulated by the runtime system, for example the list of threads of a domain, etc.) is stored off-heap. But the `config` block of memprof, and its float fields, are owned by the OCaml GC, so a-priori they cannot be accessed safely during garbage-collection, because the garbage-collector does not work per-domain, it acts on the whole heap at once. (Many parts of the GC are per-domain, but not all.) We have a situation where some parts of the garbage-collection logic access a value to make decisions while another part of the garbage-collector moves it around at the same time.

> I have yet to understand why this write in the free list ends up racing with another domain.

There are two domains at play here, one (T1) reads the memprof state, the other (T2) promotes the floating-point value that is part of the memprof state. When I reproduced locally I could observe two different race reports depending on the timing of both threads:
- if T1 reads the state first, T2 later performs a racy write by promoting the float (and writing a forwarding pointer)
- if T2 promotes the value first, the config field is updated in-place and T1 performs a racy read into memory being initialized by T2

> Indeed they can [be moved by compaction]

But the compaction code does not necessarily access the memprof runtime state, so it is possible that the present PR solves the problem for memprof. It does at least avoid the race that we are seeing reported, with races caused by the promotion logic -- by ensuring that the whole state starts in the major heap, as was already done for the `config` block.

#1306 may be a more robust solution, it ensures that the GC does not access the memprof state concurrently with garbage-collection by other domains -- the other domain must be finishing their domain-local work, or running mutator code. One way to see the change is that we reserve memprof access to only dynamic sections that are safe for mutator code. But it is more invasive -- it changes the whole GC logic in a way I am not sure is correct -- so I would be in favor of merging this one first, if we can agree that it is correct.

(It may be that the best long-term fix would be to store the `memprof` runtime state off-heap, which I have not tried to implement.)

**Comment by stedolan** [comment]
Looking at the two proposed fixes, a third option occurs to me, and I'm wondering whether you've considered it. Memprof maintains both an on-heap `Config` structure (holding, in particular, data necessary to perform callbacks such as the callback closures), and off-heap `memprof_domain_s` and `memprof_thread_s` structures holding data necessary to schedule new samples and track active samples.

My interpretation of the bug here is that the state necessary to compute `sampling(domain)` during `caml_memprof_set_trigger` is incorrectly stored in `Config` when it should live in `memprof_domain_s`, because it is accessed at a time when heap access is tricky. Having you considered moving these fields? (Maybe this is what you meant by your "best long-term fix" comment?)

**Comment by gasche** [comment]
I had a bit of a tunnel vision I think (not being familiar at all with the new memprof internals). I am happy to try the approach that you suggest, but I would be in favor of merging the present PR first to make the TSan CI green again, if you agree that it looks fine. (I'm not *completely* sure that this is a valid way to copy a float into the major heap, but you would know if this was wrong!)

**Comment by stedolan** [comment]
The code to copy the float to the major heap looks correct to me, but your comment worries me a bit:
> (Well, they can in fact be moved around during compaction.)

Is this patch definitely enough to fix the bug, or are we now relying on compaction being rare?

(In any case I'm happy to merge this to unbreak CI and follow up with a more robust fix)

**Comment by gasche** [comment]
One positive way to look at this is that bad interaction between compaction and `memprof` would be a *different* bug (than bad interaction between minor-GC and memprof).

The immediate (rather than long-term) question is whether the compaction logic happens to rely on the memprof state in a dangerous way. There is a call to `caml_memprof_scan_roots` in the middle of the "second phase" of compaction, but skimming the memprof.c codebase suggests that this does not call the `sampling(domain)` predicate, so maybe this is fine for now.

Note: The `config` value in memprof.c seems to be used only (1) on explicit user actions (starting/stopping), (2) on value lifecycle events in mutator code, and (3) by `memprof_set_trigger`. The latter is the cause of the race because it is used in the middle of the minor GC critical section, but otherwise in seems to be reserved to sampling events as well. So maybe #14306 (which ensures that `memprof_set_trigger` is not called in GC critical sections) suffices as a long-term fix.

**Comment by NickBarnes** [comment]
I think this is OK, and fixes my bug (good catch, TSAN!). If we had mixed blocks we could keep the floats inline in the config block.

Compaction is synchronous with barriers before and after, so the writes (moves) in compaction cannot race with reads elsewhere.
My motivation for keeping the config block on the heap was to avoid having to track its liveness, as it is shared between domains, threads, etc.

**Comment by NickBarnes** [comment]
Tempted to change the `Min_lambda` stuff, maybe by adding another `CONFIG_STATUS_` value: in `sampling()` we only want to test the special case of lambda being so small as to be effectively zero.

**Comment by gasche** [comment]
> Compaction is synchronous with barriers before and after, so the writes (moves) in compaction cannot race with reads elsewhere.

There could be a race with a call to `sampling(...)` by the runtime within the compaction logic -- but there is no such call today and there is no clear reason why one would be added later.

**Comment by OlivierNicole** [comment]
I’m a bit lost here because manipulating OCaml heap values from a C function has never been forbidden, so long as the necessary precautions are taken to prevent the GC from moving these values under our feet. Is there a reason why the solution being preferred is to move these values off-heap, rather than take the usual precautions?

**Comment by gasche** [comment]
The problem is that the `config` value and its field is accessed *by* the minor-GC code. Usual FFI code has to register roots, but it is not allowed to run concurrently with the minor GC critical section; it is considered "mutator code" from the GC perspective.

#14306 changes the place where the runtime uses the `config` value to be a place where mutator code can safely run, because it cannot happen concurrently with minor-collection work from other domains.

(Note: I am not sure why @NickBarnes closed this PR, it does not look intentional.)

**Comment by NickBarnes** [comment]
I don't know how I closed this!

**Comment by gasche** [comment]
I pushed a Changes entry.

I sense some consensus to merge this to un-block the TSan CI, and then discuss a potential long-term fix at its own pace, and/but this PR needs an explicit approval to be merged.

**Review by OlivierNicole — APPROVED** [review]
The change as proposed looks correct to me.

**Comment by gasche** [comment]
Thanks! I merge now to unblock the TSan CI, and we can discuss more long-term options separately. My current best proposal is #14306, but I might explore @stedolan's suggestion in https://github.com/ocaml/ocaml/pull/14304#issuecomment-3425680451 . (This could be something to discuss at an upcoming runtime meeting.)

File: runtime/memprof.c [hunk 1]
```diff
@@ -2233,7 +2233,11 @@ CAMLprim value caml_memprof_start(value lv, value szv, value tracker)
     one_log1m_lambda = MIN_ONE_LOG1M_LAMBDA; /* negative infinity */
   }
 
-  one_log1m_lambda_v = caml_copy_double(one_log1m_lambda);
+  lambda_v = caml_alloc_shr(Double_wosize, Double_tag);
+  Store_double_val(lambda_v, lambda);
```
Comment (jhjourdan) [inline runtime/memprof.c:2237]: Maybe this remark is silly, but still: what is the point of allocating `lambda_v` (be it in the major or in the minor heap) if we never use it after all? Before the patch, `lambda_v` did not exist.

Comment (jhjourdan) [reply runtime/memprof.c:2237]: You probably want to use `lambda_v` instead of `lv` below.

Comment (gasche) [reply runtime/memprof.c:2237]: Indeed, good catch! I will fix by direct push to trunk.

```
  caml_initialize(&Field(config, CONFIG_FIELD_LAMBDA), lv);
```

should be

```
  caml_initialize(&Field(config, CONFIG_FIELD_LAMBDA), lambda_v);
```

(maybe I should have used shadowing creatively to avoid this issue)

**Comment by gasche** [comment]
@jhjourdan caught a small bug in this PR which I fixed in trunk in 14980901e121ea58bfac2574a8ac1ffc0ced0fc8.

**Comment by jhjourdan** [comment]
> @jhjourdan caught a small bug in this PR which I fixed in trunk in https://github.com/ocaml/ocaml/commit/14980901e121ea58bfac2574a8ac1ffc0ced0fc8.

.... Which in fact was not a bug, because the field `CONFIG_FIELD_LAMBDA` is in fact never read in the runtime. There is no need to allocate `lambda_v`.

**Comment by gasche** [comment]
Indeed, moving `one_log1m_lambda` to the major heap suffices to fix the TSan-reported issue, but I prefer to play it safe by ensuring that the whole configuration is on the major heap, instead of handling different fields inconsistently.

**Comment by NickBarnes** [comment]
> > @jhjourdan caught a small bug in this PR which I fixed in trunk in [1498090](https://github.com/ocaml/ocaml/commit/14980901e121ea58bfac2574a8ac1ffc0ced0fc8).
> 
> .... Which in fact was not a bug, because the field `CONFIG_FIELD_LAMBDA` is in fact never read in the runtime. There is no need to allocate `lambda_v`.

FYI the config has a slot for the `lambda` value purely for debugging purposes; at some point in development I had a Caml function which would return it, and logging code which included it in messages.

**Comment by jhjourdan** [comment]
Yeah, I got that.

I guess we could now remove this field. If debugging were necessary, the `1log1ml` field contains enough info to retreive the value of `lambda`, modulo some rounding errors.
